File: src/com/gahon/medium/_002/AddTwoNumbers.java
package com.gahon.medium._002;

/**
 * Definition for singly-linked list.
 */
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}


public class AddTwoNumbers {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode(0), cur = result;
        int flag = 0;
        for (; l1 != null || l2 != null || flag > 0; ) {
            int n1 = l1 == null ? 0 : l1.val;
            int n2 = l2 == null ? 0 : l2.val;
            int sum = (n1 + n2 + flag);
            l1 = l1!=null?l1.next:null;
            l2 = l2!=null?l2.next:null;
            flag = sum / 10;
            ListNode node = new ListNode(sum%10);
            cur.next = node;
            cur = node;
        }
        return result.next;
    }

}

File: solution/C++/002.Add_Two_Numbers/addTwoNumbers.cpp
/**
 * Definition for singly-linked list.
 */
 struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode *res = new ListNode(0),*newres =res;
        ListNode *longList;
        while((l1!=NULL)&&l2!=NULL){
        	ListNode *tmp = new ListNode(0);
        	carry = l1->val+l2->val+carry;
        	tmp->val = carry%10;
        	carry = carry/10;
        	l1 = l1->next;
        	l2 = l2->next;

        	newres->next = tmp;
        	newres = tmp;
        }
        longList = (l1!=NULL)?l1:l2;
        while(longList!=NULL){
        	ListNode *tmp = new ListNode(0);
        	carry = longList->val+carry;
        	tmp->val = carry%10;
        	carry = carry/10;
        	longList=longList->next;
        	newres->next = tmp;
        	newres = tmp;
        }
        if(carry==1){
        	ListNode *tmp = new ListNode(1);
        	newres->next = tmp;
        }
        return res->next;
    }
};